### ISSUE TITLE
Error deserializing subclassed field

### ISSUE THREAD

**Issue by Slimu** [msgpack/msgpack-java#124]
Hello, I've stumbled upon a problem when trying to deserialize a message. I have a base class A which is extended by class B. I serialize a class C which has one of the fields of type A. But this field is set with the value of a class B variable. At deserialization only the fields from A are populated and if I'm using instanceof to test the class of the field, is of type A not B.

Is there any way to indicate that at deserialization, to create an instance of a subclass? Maybe a custom field that would indicate the expected class of the field?

**Comment by komamitsu** [comment]
@Slimu Sorry for late reply. In that case, you need to write a custom template class that extends AbstractTemplate<A> and switches the way of serialization/deserialization in `writeTo()` and `readFrom()` methods depending on which class a target object is A or B.

BTW, you can deal with this sort of issues easily using https://github.com/msgpack/msgpack-java/tree/v07-develop/msgpack-jackson in v0.7.x like this.

```
    @JsonTypeInfo(use=JsonTypeInfo.Id.NAME, include=JsonTypeInfo.As.PROPERTY, property="objectType")
    @JsonSubTypes({
            @JsonSubTypes.Type(value=B.class)
    })
    public static class A {
        public int x;
    }

    @JsonTypeInfo(use=JsonTypeInfo.Id.NAME, include=JsonTypeInfo.As.PROPERTY, property="objectType")
    public static class B extends A {
        public int y;
    }

    public static class C {
        public A a;
    }

    public static void main(String args[]) {
            ObjectMapper mapper = new ObjectMapper(new MessagePackFactory());
            B a = new B();
            a.x = 1;
            a.y = 2;
            C c = new C();
            c.a = a;
            ByteArrayOutputStream out = new ByteArrayOutputStream();
            mapper.writeValue(out, c);
            C restored = mapper.readValue(out.toByteArray(), C.class);
    }
```

**Comment by Slimu** [comment]
@komamitsu Thank you for your response! Since when I posted the issue I fixed the problem in the following way:
For class C, I've written custom writeTo() and readFrom() methods. When serializing an A/B object, I write the canonical name of the class after which I write the object (or null instead if the object is null). During deserialization, I check for the className String, and if I have it, I use something like (A) unpacker.read(Class.forName(className))
